### PR TITLE
Provider Airbyte: breaking change, update provider to use Airbyte API Python SDK

### LINKED ISSUE #41365: [Providers] Airbyte new deployment version changed API endpoints
### Apache Airflow Provider(s)

airbyte

### Versions of Apache Airflow Providers

apache-airflow-providers-airbyte==3.5.1

### Apache Airflow version

2.8.1

### Operating System

MWAA/AL

### Deployment

Amazon (AWS) MWAA

### Deployment details

Standard MWAA deployment. Nothing really specific or particular

### What happened

After updating to one of the latest Airbyte versions (and using acbtl installation method, as they have as standard on their documentation) the Airbyte provider from Airflow cannot connect to the api server. The reason is that the endpoints have changed, according to Airbyte's documentation:

https://github.com/airbytehq/airbyte/issues/43423

https://reference.airbyte.com/reference/standalone-server-deprecation-and-migration-to-airbyte-server

### What you think should happen instead

Provider should be updated so that it points to the correct endpoint of the new api for Airbyte server

### How to reproduce

Deploy MWAA. Perform a fresh install of Airbyte using current documentation and standards for OSS installation (abctl).
Use Airflow's Airbyte provider to do anything in the Airbyte server and you will get a connection refused error, as it is pointing to the wrong endpoint.

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by marcosmarxm**
Hello 👋, Airbyte has finally consolidated API access across the open-source, cloud, and enterprise versions. 
This means all of them have a single method `client_id` and `client_secret` to [authenticate to the API](https://reference.airbyte.com/reference/authentication) now. 
This pull request remove the dependency of `HttpHook` and uses the `airbyte-api` Python SDK library to execute all operations (create job, cancel job, etc). 
The change is a breaking change to users as it removed the `api_type` parameter as it isn't required anymore. 
The new authentication method will become soon the unique method to auth to the API so this is a required changed.
Unit tests were updated to run using the new code.

---

### CODE REVIEW

File: airflow/providers/airbyte/hooks/airbyte.py
Comment (marcosmarxm) [inline airflow/providers/airbyte/hooks/airbyte.py:59]: Removed this to start using the enum class provided by the Airbyte API.

File: airflow/providers/airbyte/sensors/airbyte.py
Comment (marcosmarxm) [inline airflow/providers/airbyte/sensors/airbyte.py]: I encountered a few errors while updating the tests for sensors requesting to add the `skip_policy`. Since I don't use Airflow frequently and don't contribute much, I need help to ensure if this is.

File: docs/apache-airflow-providers-airbyte/operators/airbyte.rst
Comment (eladkal) [inline docs/apache-airflow-providers-airbyte/operators/airbyte.rst]:
```suggestion
Airbyte offers a single method to authenticate for Cloud and OSS users.
```

File: airflow/providers/airbyte/provider.yaml
Comment (eladkal) [inline airflow/providers/airbyte/provider.yaml]: Do we still need http provider as dependency if we move to sdk?

File: airflow/providers/airbyte/provider.yaml
Comment (eladkal) [inline airflow/providers/airbyte/provider.yaml:1]: Please add `4.0.0` under `versions:`

File: airflow/providers/airbyte/hooks/airbyte.py
Comment (eladkal) [reply airflow/providers/airbyte/hooks/airbyte.py:59]: Please add entry of 4.0.0 in the changelog of the provider with 1-2 paragraph explaining how users should migrate from previous version to the current one.

**Review by eladkal — CHANGES_REQUESTED** [review]

File: airflow/providers/airbyte/provider.yaml
Comment (marcosmarxm) [reply airflow/providers/airbyte/provider.yaml]: No, nice catch!

**Comment by marcosmarxm** [comment]
@eladkal, can you take another look? Some tests are failing, but I can't determine if it’s due to my changes.

**Comment by potiuk** [comment]
If in doubt - rebase:

Screenshot 2024-08-05 at 15 20 56

File: airflow/providers/airbyte/hooks/airbyte.py
Comment (Imbruced) [inline airflow/providers/airbyte/hooks/airbyte.py]: nit: here you pass exception directly 
https://github.com/apache/airflow/pull/41122/files#diff-f4527df370ba3e14e4d474da1a3698416349f915d96537f13856e59b86b71230R104

File: airflow/providers/airbyte/hooks/airbyte.py
Comment (Imbruced) [inline airflow/providers/airbyte/hooks/airbyte.py]: nit: similar like above https://github.com/apache/airflow/pull/41122/files#diff-f4527df370ba3e14e4d474da1a3698416349f915d96537f13856e59b86b71230R104

File: airflow/providers/airbyte/operators/airbyte.py
Comment (Imbruced) [inline airflow/providers/airbyte/operators/airbyte.py:86]: nit: what do you think about keeping happy path on the left with early return ? 
```python
  def execute(self, context: Context) -> None:
        """Create Airbyte Job and wait to finish."""
        hook = AirbyteHook(airbyte_conn_id=self.airbyte_conn_id, api_version=self.api_version)
        job_object = hook.submit_sync_connection(connection_id=self.connection_id)
        self.job_id = job_object.job_id
        state = job_object.status
        end_time = time.time() + self.timeout

        self.log.info("Job %s was submitted to Airbyte Server", self.job_id)
        if not self.deferrable:
            hook.wait_for_job(job_id=self.job_id, wait_seconds=self.wait_seconds, timeout=self.timeout)
            
        if not self.deferrable or self.asynchronous:
            self.log.info("Job %s completed successfully", self.job_id)
            return self.job_id
        
        if state in (JobStatusEnum.RUNNING, JobStatusEnum.PENDING, JobStatusEnum.INCOMPLETE):
            self.defer(
                timeout=self.execution_timeout,
                trigger=AirbyteSyncTrigger(
                    conn_id=self.airbyte_conn_id,
                    job_id=self.job_id,
                    end_time=end_time,
                    poll_interval=60,
                ),
                method_name="execute_complete",
            )
        elif state == JobStatusEnum.SUCCEEDED:
            self.log.info("Job %s completed successfully", self.job_id)
            return
        elif state == JobStatusEnum.FAILED:
            raise AirflowException(f"Job failed:\n{self.job_id}")
        elif state == JobStatusEnum.CANCELLED:
            raise AirflowException(f"Job was cancelled:\n{self.job_id}")
        else:
            raise AirflowException(
                f"Encountered unexpected state `{state}` for job_id `{self.job_id}"
            )
```

Comment (marcosmarxm) [reply airflow/providers/airbyte/operators/airbyte.py:86]: Awesome! Thanks for the suggestion @Imbruced function looks way easier to understand.

File: airflow/providers/airbyte/CHANGELOG.rst
Comment (eladkal) [inline airflow/providers/airbyte/CHANGELOG.rst]:
```suggestion
```

This is auto generated by release process

**Comment by marcosmarxm** [comment]
Before merging this change I'll run a final test with Airbye cloud and open source.

**Comment by eladkal** [comment]
> Before merging this change I'll run a final test with Airbye cloud and open source.

Please do. If no further comments we are good to go

**Comment by potiuk** [comment]
You need to rebase/run pre-commit

**Comment by marcosmarxm** [comment]
Running final test with local deployment and cloud. Gonna 💚 🚀  after finish these tests.

**Comment by marcosmarxm** [comment]
🟢 Tests are working
🟢 I executed DAGs with a local deployment and cloud, both are working.

Need rebase and 🚀

**Comment by marcosmarxm** [comment]
@eladkal can you help figure out the problem with test running `breeze release-management prepare-provider-documentation --include-not-ready-providers --non-interactive`

**Comment by potiuk** [comment]
> @eladkal can you help figure out the problem with test running breeze release-management prepare-provider-documentation --include-not-ready-providers --non-interactive

The problem is that we are changing version of Airbyte while the RC candidate for the current reelase is not yet released. That's a small flaw we have in the process which we might want to fix soon - but this one will get fixed right after @eladkal will release the latest batch of providers (should be soon).

**Comment by potiuk** [comment]
BTW. Generally speaking it is correct - we are not able to prepare documentation for new version as we find an inconsistency between what is stored as previous relase and what TAGs we have in Airflow, so technically speaking this error is pretty "valid" - but possibly we might want to skip it in CI @amoghrajesh -> maybe you can take a look at it? 

The error is that we are trying to get changes since last release:

```
 git log '--pretty=format:%H %h %cd %s' --date=short apache-https-for-providers/main...providers-airbyte/3.9.0 -- .
  
  ========================= STDERR start ============================
  fatal: bad revision 'apache-https-for-providers/main...providers-airbyte/3.9.0'
  
  ========================= STDERR end ==============================
```

And one of the ways to avoid it is to handle it and see if there is a latest `rc` that we can use instead.

**Comment by potiuk** [comment]
Or at a very least provide a better message in this case with the note "please wait until the release candidate for previous version is released".

**Comment by marcosmarxm** [comment]
Thanks for the update @potiuk I kinda saw after searching for the airbyte-provider in pypi and checking the latest version if 3.8.X. In this situation we need to wait to have the new release to after get this merged, right?

**Comment by potiuk** [comment]
> In this situation we need to wait to have the new release to after get this merged, right?

Yes. We have provider waves every 2-4 weeks - just the last one was "big" because we were bumping min version of all providers.

**Comment by marcosmarxm** [comment]
@eladkal I think the single failing test isn't related to my change. Is there anything else I need to do? Thanks for all support

**Comment by marcosmarxm** [comment]
@potiuk / @eladkal it looks tests are almost all green. There is only 1 test failing because of time out but similar tests are green. Anything else I need to do? The branch was rebased with latest main.

**Review by eladkal — APPROVED** [review]

**Comment by baugarcia** [comment]
Great! any ETA on when will the new version be deployed in the python repository? Version is still 3.9.0 there: 
https://pypi.org/project/apache-airflow-providers-airbyte/

**Comment by marcosmarxm** [comment]
@baugarcia probably the next window of releases is in 3~4 weeks. But keep this to the maintainers to provide a better answer.

**Pull request merged**
